connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1020,7 +1107,9 @@ public HerderRequest(long at, Callable<Void> action, Callback<Void> callback) {
 
         @Override
         public int compareTo(HerderRequest o) {
-            return Long.compare(at, o.at);
+            final int soonest = Long.compare(at, o.at);
+            // If tied, returning a positive value should respect insertion order.
+            return soonest != 0 ? soonest : 1;
```

Comment:
It feels arbitrary, but actually corresponds to what `PriorityQueue` does in the presence of duplicates. If an element is inserted in the PQ that has the same timestamp (field `at` here) with an element already in the queue, then the new element will be placed later in the queue (according to their order of the PQ). 

`ConcurrentSkipListSet ` is a set, and therefore it does not allow duplicates. Without breaking the tie in `compareTo` the new element would override an older element with the same priority. However, the way I break this tie, I simulate what a PriorityQueue would do. The new element will be placed later. Thus, it preserves the insertion order. To see why this happens, you can check how `compareTo` is used within `ConcurrentSkipListSet`'s impelementation (method `doPut` in `ConcurrentSkipListMap` actually).